File: kotlin/0948-bag-of-tokens.kt
class Solution {
    fun bagOfTokensScore(tokens: IntArray, power: Int): Int {
        tokens.sort()

        var res = 0
        var score = 0
        var power = power

        var left = 0
        var right = tokens.lastIndex
        while (left <= right) {
            if (power >= tokens[left]) {
                power -= tokens[left++]
                score++
                res = maxOf(res, score)
            } else if (score > 0) {
                power += tokens[right--]
                score--
            } else {
                break
            }
        }

        return res
    }
}

File: go/0948-bag-of-tokens.go
func bagOfTokensScore(tokens []int, power int) int {
	res, score := 0, 0
	sort.Ints(tokens)

	l, r := 0, len(tokens)-1
	for l <= r {
		if power >= tokens[l] {
			power -= tokens[l]
			l += 1
			score += 1
			res = max(res, score)
		} else if score > 0 {
			power += tokens[r]
			r -= 1
			score -= 1
		} else {
			break
		}
	}
	return res
}
